### PR TITLE
Do not allow IdentityItems with empty ids (resolves #100)

### LINKED ISSUE #100: IdentityMap should not accept invalid identity items with empty ids
### Task description

Clean up/ignore identity items with empty string ids when updating the identityMap (update/remove).
Clean up previously persisted identity items with empty ids.

### Additional implementation details or code snippet(s)

_No response_

### PR DESCRIPTION
**Pull request by kevinlind**
## Description

Do not allow IdentityItems with empty ids.

- If an IdentityItem is created with an empty or null id, throw IllegalArgumentException from constructor.
- Add tests to validate reading JSON Identity Maps which contain items with null/empty ids are correctly removed when the IdentityMap object is created.

See Issue #100 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java [hunk 1]
```diff
@@ -49,8 +50,8 @@ public IdentityItem(
 		@Nullable final AuthenticatedState authenticatedState,
 		final boolean primary
 	) {
```
Comment (addb) [inline code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:52]: Update doc:
`@throws IllegalArgumentException if id is null or empty`

File: code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java [hunk 2]
```diff
@@ -193,7 +194,15 @@ static IdentityItem fromData(final Map<String, Object> data) {
 			return new IdentityItem(id, authenticatedState, primary);
```
Comment (addb) [inline code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:194]: Update doc:
`@throws IllegalArgumentException if id is null or empty`

**Review by addb — APPROVED** [review]
Left 1 comment for doc update. Apart from that looks good!

File: code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java [hunk 3]
```diff
@@ -42,15 +43,15 @@ public final class IdentityItem {
...
 	 */
 	public IdentityItem(
 		@NonNull final String id,
 		@Nullable final AuthenticatedState authenticatedState,
 		final boolean primary
 	) {
-		if (id == null) {
-			throw new IllegalArgumentException("id must be non-null");
+		if (StringUtils.isNullOrEmpty(id)) {
+			throw new IllegalArgumentException("id must be non-null and non-empty");
```
Comment (emdobrin) [inline code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:54]: is there a way to avoid throwing an exception for empty id value? I am thinking this could potentially break an existing implementation which erroneously sends empty ids

File: code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityMapTests.java [hunk 4]
```diff
@@ -310,6 +310,83 @@ public void test_FromData() throws Exception {
 		assertEquals("false", flattenedMap.get("identityMap.USERID[0].primary"));
 	}
 
+	@Test
+	public void test_FromData_removesItem_emptyId() throws Exception {
```
Comment (emdobrin) [inline code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityMapTests.java:314]: Are we also covering the use-case where there may be previously stored ids with empty id value, so that the SDK clears them and stores the updated identity map? And if so, do we need a test for this scenario?

File: code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java [hunk 5]
```diff
@@ -42,15 +43,15 @@ public final class IdentityItem {
 	 * @param id                 id for the item; should not be null
```
Comment (emdobrin) [inline code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:43]: id for the item; should not be null or empty 

Also, don't forget to update the api reference page if we don't already cover this

**Comment by kevinlind** [comment]
Hi @addb and @emdobrin . I've updated to code to ignore IdentityItems with empty ids instead of throwing an exception. Changing the IdentityItem API to throw is a breaking change as Emilia pointed out. I've added some functional tests to validate existing items with empty ids are not loaded on bootup. I also updated the documentation to state IdentityItems with empty ids are ignored when adding to an IdentityMap.

**Comment by addb** [comment]
> Hi @addb and @emdobrin . I've updated to code to ignore IdentityItems with empty ids instead of throwing an exception. Changing the IdentityItem API to throw is a breaking change as Emilia pointed out. I've added some functional tests to validate existing items with empty ids are not loaded on bootup. I also updated the documentation to state IdentityItems with empty ids are ignored when adding to an IdentityMap.

I am not sure if this is the right approach as now we have 3 behaviors
|id|outcome|
|--|--|
|null | throws exception|
| emty| ignores|
|non-empty non-null String| accepted|

We should have similar behavior for error case, either we throw exceptions for both or ignore both

File: Documentation/api-reference.md [hunk 6]
```diff
@@ -432,7 +432,7 @@ Identity.updateIdentities(identityMap)
 
 ### IdentityMap
 
-Defines a map containing a set of end user identities, keyed on either namespace integration code or the namespace ID of the identity. The values of the map are an array, meaning that more than one identity of each namespace may be carried.
+Defines a map containing a set of end user identities, keyed on either namespace integration code or the namespace ID of the identity. The values of the map are an array, meaning that more than one identity of each namespace may be carried. Identities, represented by [IdentityItem](#identityitem) objects, may not have null or empty identitifers and are ignored when adding to the IdentityMap.
```
Comment (emdobrin) [inline Documentation/api-reference.md:435]: Defines a map containing a set of end user identities, keyed on either namespace integration code or the namespace ID of the identity. The values of the map are an array of [IdentityItem](#identityitem)(s), meaning that more than one identity of each namespace may be carried. Each IdentityItem should have a valid, non-null and non-empty identifier, otherwise it will be ignored.

File: code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java [hunk 7]
```diff
@@ -64,6 +67,7 @@ public IdentityItem(
 	 * (@code primary} is set to false
 	 *
 	 * @param id the id for this {@link IdentityItem}; should not be null
+	 * @throws IllegalArgumentException if id is null or empty
```
Comment (emdobrin) [inline code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:70]: if {@code id} is null

**Review by emdobrin — APPROVED** [review]

File: code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java [hunk 8]
```diff
@@ -37,12 +37,15 @@ public final class IdentityItem {
...
+	 *
 	 *
 	 * @param id                 id for the item; should not be null
 	 * @param authenticatedState {@link AuthenticatedState} for the item; if none is provided {@link AuthenticatedState#AMBIGUOUS} is used as default
 	 * @param primary            primary flag for the item
-	 * @throws IllegalArgumentException if id is null
+	 * @throws IllegalArgumentException if {@code id} is null
 	 */
 	public IdentityItem(
 		@NonNull final String id,
```
Comment (addb) [inline code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:51]: `@NotBlank`
https://javaee.github.io/javaee-spec/javadocs/javax/validation/constraints/NotBlank.html

File: code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java [hunk 9]
```diff
@@ -64,6 +67,7 @@ public IdentityItem(
 	 * (@code primary} is set to false
 	 *
 	 * @param id the id for this {@link IdentityItem}; should not be null
+	 * @throws IllegalArgumentException if id is null or empty
 	 */
 	public IdentityItem(@NonNull final String id) {
```
Comment (addb) [inline code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:72]: @NotBlank
https://javaee.github.io/javaee-spec/javadocs/javax/validation/constraints/NotBlank.html

File: code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java (hunk 8 above)
Comment (kevinlind) [reply code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityItem.java:51]: This is not a standard AndroidX annotation, but instead from Hibernate-Validator in Java EE, so it's not really available to us.
